torch/_inductor/ir.py

```diff
@@ -2872,13 +2873,29 @@ def create(cls, inputs, dim):
             inputs=[],
         )
         kernel = StorageBox(concat_kernel)
+        buffer_names = []
         for i in range(len(inputs)):
-            kernel.data.inputs.append(
-                cls.realize_into(
-                    inputs[i],
-                    SliceView.create(kernel, dim, offsets_start[i], offsets_end[i]),
-                )
+            input_buffer = cls.realize_into(
+                inputs[i],
+                SliceView.create(kernel, dim, offsets_start[i], offsets_end[i]),
             )
+
+            kernel.data.inputs.append(input_buffer)
+            if isinstance(inputs[i].data, BaseView):
+                input_unwrapped = inputs[i].data.unwrap_view()
+            else:
+                input_unwrapped = inputs[i].data
+
+            if (
+                input_unwrapped.is_input_buffer()
```

Comment:
Do I get it right that the horizontal fusion of separate concat kernels into foreach is enabled only when the concat arguments are graph inputs? I'm wondering if this could be extended to the cases when the arguments are intermediate (realized) buffers in the graph. Thanks!

Reply:
Yeah this is true, I'm curious about your use case because I think in most cases of intermediates the concat kernel rewrites the storage of the upstream ops to point to the concat buffer rather than launching a copy kernel. Let me know if there are cases where we copy anyway and I can remove this constraint.

Reply:
Thanks for the explanation. In my use case, concat takes (many) arguments which are outputs of the `torch.ops.aten.slice.Tensor` ops. So my understanding is that the latter are already views (reading from a particular offset of the input to the slice ops). At the same time, the slices / views are not graph inputs. So the `ConcatKernel` actually ends up generating a bunch of copy kernels: one per each slice input.

Reply:
I see, yeah I'll experiment with removing this constraint in a separate PR since this has had a lot of edge cases and I want the MRS folks to test this. It should just work ootb I think